src/components/atoms/Timeline/index.tsx

```diff
@@ -0,0 +1,244 @@
+import { XXSRegular } from "@reearth/theme/fonts";
+
+type Range = {
+  start: number;
+  end: number;
+};
+
+export type Props = {
+  /**
+   * @description
+   * This value need to be epoch time.
+   */
+  currentTime: number;
+  /**
+   * @description
+   * These value need to be epoch time.
+   */
+  range?: { [K in keyof Range]?: Range[K] };
+  renderKnob: () => ReactElement;
+  knobSize: number;
+};
+
+const EPOCH_SEC = 1000;
+const PADDING_HORIZONTAL = metricsSizes["m"];
+const STRONG_MEMORY_HOURS = 3;
+const STRONG_MEMORY_WIDTH = 2;
+const NORMAL_MEMORY_WIDTH = 1;
+const GAP_HORIZONTAL = 14;
+const HOURS_SECS = 3600;
+const MEMORY_INTERVAL = 600;
+const DAY_SECS = 86400;
+
+const getRange = (range: Props["range"]): Range => {
+  const { start, end } = range || {};
+  if (start !== undefined && end !== undefined) {
+    return { start, end };
+  }
+
+  if (start !== undefined && end === undefined) {
+    return {
+      start,
+      end: start + DAY_SECS,
+    };
+  }
+
+  if (start === undefined && end !== undefined) {
+    return {
+      start: Math.max(end - DAY_SECS, 0),
+      end,
+    };
+  }
+
+  const defaultStart = Date.now();
+
+  return {
+    start: defaultStart,
+    end: defaultStart + DAY_SECS * EPOCH_SEC,
+  };
+};
+
+const Timeline: React.FC<Props> = memo(
+  function TimelinePresenter({ currentTime, range: _range, renderKnob, knobSize }) {
+    const range = useMemo(() => {
+      const range = getRange(_range);
+      if (process.env.NODE_ENV !== "production") {
+        if (range.start > range.end) {
+          throw new Error("Out of range error. `range.start` should be less than `range.end`");
+        }
+      }
+      return range;
+    }, [_range]);
+    const { start, end } = range;
+    const epochDiff = end - start;
+
+    // convert epoch diff to second.
+    const memoryCount = useMemo(
+      () => Math.trunc(epochDiff / EPOCH_SEC / MEMORY_INTERVAL),
+      [epochDiff],
+    );
+
+    const startDate = useMemo(() => new Date(start), [start]);
+    const hoursCount = Math.trunc(HOURS_SECS / MEMORY_INTERVAL);
+    // Convert memory count to pixel.
+    const currentPosition = useMemo(() => {
+      const diff = Math.min((currentTime - start) / EPOCH_SEC / MEMORY_INTERVAL, memoryCount);
+      const strongMemoryCount = diff / (hoursCount * STRONG_MEMORY_HOURS);
+      return Math.max(
+        diff * GAP_HORIZONTAL +
+          (diff - strongMemoryCount) * NORMAL_MEMORY_WIDTH +
+          strongMemoryCount * STRONG_MEMORY_WIDTH +
+          STRONG_MEMORY_WIDTH / 2,
+        0,
+      );
+    }, [currentTime, start, memoryCount, hoursCount]);
+
+    return (
+      <Container>
+        <MemoryBox>
+          <MemoryList start={startDate} memoryCount={memoryCount} hoursCount={hoursCount} />
+          <Icon
+            style={{
+              left: currentPosition + PADDING_HORIZONTAL - knobSize / 2,
+            }}>
+            {renderKnob()}
+          </Icon>
```

Comment:
Sorry I thought I left a comment here, so if this is a repeat, I apologize.
But is there a reason you are doing the Icon like this? Does the icon need to be fully customizable (even to the point of not using the Icon atom component)?

Reply:
👍 [fix: remove renderKnob props](https://github.com/reearth/reearth-web/pull/256/commits/186197e65cf71dda4b0a2b49e36b4999d8c8a835)